### PR TITLE
[BDHK-206] Implement whitelisting for addresses that can execute tx onchain

### PR DESCRIPTION
**Pull request by adelowo**
- Added env value to store whitelisted addresses
- Use badger as a kv store for these addresses ( faster than iterating over them for every user op in a batch )
- Added tracing

This will be enabled implicitly if any whitelisted address is added as below.

```
ERC4337_BUNDLER_ADDRESS_WHITELIST=0xaddress1
```

An empty value disables the middleware that checks for whitelisting from running

### CODE REVIEW

File: srv/whitelist.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+			if err != nil {
+				logger.Error(err, "sender not found in whitelist. cannot execute transaction",
+					"address", userop.Sender.Hex())
+				return err
+			}
+		}
+
+		return nil
+	}
+}
```
Comment (hpsing) [inline srv/whitelist.go]:
```suggestion
func CheckSenderWhitelist(db *badger.DB, 
whitelistedAddresses []common.Address, 
logger logr.Logger) modules.BatchHandlerFunc {
	logger.Info("Setting up addresses in BadgerDB for whitelist checks")
	if len(whitelistedAddresses) > 0 {
		logger.Info("Bundler address whitelisting enabled", "number_of_whitelisted_addresses", len(whitelistedAddresses))
	}

	// Use a single transaction to store all addresses
	err := db.Update(func(txn *badger.Txn) error {
		for _, addr := range whitelistedAddresses {
			if err := txn.Set([]byte(getWhitelistKey(addr)), addr.Bytes()); err != nil {
				logger.Error(err, "Failed to store address in DB for whitelisting", "address", addr.Hex())
				return err
			}
			logger.Info("Address stored in DB whitelist", "address", addr.Hex())
		}
		return nil
	})

	if err != nil {
		logger.Error(err, "Failed to store multiple addresses in BadgerDB")
		return nil
	}

	return func(ctx *modules.BatchHandlerCtx) error {

		_, span := utils.GetTracer().
			Start(context.Background(), "WithAdressWhitelist")
		defer span.End()

		if len(whitelistedAddresses) == 0 {
			logger.Info("Address whitelisting not enabled. Skipping middleware.")
			return nil
		}

		for _, userop := range ctx.Batch {
			senderKey := getWhitelistKey(userop.Sender)
			err := db.View(func(txn *badger.Txn) error {
				_, err := txn.Get([]byte(senderKey))
				return err
			})
			if err != nil {
				logger.Error(err, "Sender not found in whitelist; cannot execute transaction", "address", userop.Sender.Hex())
				return err
			}
		}
		return nil
	}
}
```

**Review by hpsing — CHANGES_REQUESTED** [review]
couple of suggestions:

1. Removed the panic on error and instead returned errors when they occurred during the DB operation.
2. Change the function to batch all whitelisted address insertions into a single BadgerDB transaction instead of separate transactions for each address.
3. Add godoc.
4. Add test cases.

File: cmd/main.go [hunk 2]
```diff
@@ -116,11 +116,18 @@ func main() {
...
+
+	if whitelistHandler == nil {
+		stdLogger.Info("could not set up sender whitelist middleware")
+		os.Exit(1)
+	}
+
 	bundlerClient.UseModules(
 		exp.DropExpired(),
 		batch.SortByNonce(),
+		whitelistHandler,
```
Comment (blewater) [inline cmd/main.go]: Why not first like an app firewall?

File: srv/whitelist.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+	"github.com/blndgs/bundler/utils"
+	"github.com/dgraph-io/badger/v3"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/go-logr/logr"
+	"github.com/stackup-wallet/stackup-bundler/pkg/modules"
+)
+
+func getWhitelistKey(addr common.Address) string {
+	return fmt.Sprintf("bundler-whitelist-%s", addr.Hex())
```
Comment (blewater) [inline srv/whitelist.go]: `bundler-` is redundant, no other component uses this badger db instance.

File: srv/whitelist_test.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+	"github.com/stackup-wallet/stackup-bundler/pkg/userop"
+	"github.com/stretchr/testify/require"
+)
+
+func TestCheckSenderWhitelist(t *testing.T) {
+
+	db, err := badger.Open(badger.DefaultOptions("testdata/test.db"))
+	if err != nil {
+		log.Fatal(err)
+	}
```
Comment (blewater) [inline srv/whitelist_test.go]: you may use a memory badger instance which does not leave artifacts.
```Go
    db, err := badger.Open(badger.DefaultOptions("").WithInMemory(true))
```

File: srv/whitelist.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+		for _, userop := range ctx.Batch {
+			senderKey := getWhitelistKey(userop.Sender)
+			err := db.View(func(txn *badger.Txn) error {
+				_, err := txn.Get([]byte(senderKey))
+				return err
+			})
+			if err != nil {
+				logger.Error(err, "Sender not found in whitelist; cannot execute transaction", "address", userop.Sender.Hex())
+				return err
+			}
```
Comment (blewater) [inline srv/whitelist.go]: Filter out any non-allowed rather than reject the batch.

File: srv/whitelist_test.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+		common.HexToAddress("0x000000000000000000000000000000000000dEaD"),
+	}, logr.Discard())
+
+	require.NoError(t, handler(&modules.BatchHandlerCtx{
+		Batch: []*userop.UserOperation{
+			{
+				Sender: common.HexToAddress("0x000000000000000000000000000000000000dEaD"),
+			},
+		},
+	}))
```
Comment (blewater) [inline srv/whitelist_test.go]: Add use case scenarios that check post-filtering batch counts:
1. Empty whitelist
2. Mixed allowed non-allowed white lists
3. All whitelisted addresses.

**Review by blewater — CHANGES_REQUESTED** [review]
Nice touch with the database! A few comments.

**Review by hpsing — APPROVED** [review]

File: srv/whitelist.go [hunk 7]
```diff
@@ -58,17 +58,18 @@ func CheckSenderWhitelist(db *badger.DB,
...
 			senderKey := getWhitelistKey(userop.Sender)
 			err := db.View(func(txn *badger.Txn) error {
 				_, err := txn.Get([]byte(senderKey))
 				return err
 			})
 			if err != nil {
-				logger.Error(err, "Sender not found in whitelist; cannot execute transaction", "address", userop.Sender.Hex())
-				return err
+				logger.Error(err, "Sender not found in whitelist; removing transaction from batch", "address", userop.Sender.Hex())
+				ctx.Batch = append(ctx.Batch[:idx], ctx.Batch[idx+1:]...)
```
Comment (blewater) [inline srv/whitelist.go]: 👍

**Review by blewater — APPROVED** [review]
Nice work!

**Pull request merged**